File: android/app/src/main/java/gg/yappy/app/data/SupportLinks.kt
package gg.yappy.app.data

import android.net.Uri
import gg.yappy.app.BuildConfig

object SupportLinks {
    /** Carry only the case capability onto our configured support site. */
    fun url(appeal: Boolean = false, source: String? = null): String {
        val builder = Uri.parse(BuildConfig.WEB_URL.trimEnd('/') + "/support/").buildUpon()
            .appendQueryParameter("client", "Android · ${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})")
        if (appeal) {
            builder.appendQueryParameter("topic", "appeal")
            val token = runCatching {
                val fragment = Uri.parse(source.orEmpty()).encodedFragment ?: return@runCatching null
                Uri.parse("https://yappy.gg/?$fragment").getQueryParameter("appeal")
            }.getOrNull()
            if (token != null && token.length <= 2000 && token.matches(Regex("[A-Za-z0-9._-]+"))) {
                builder.fragment("appeal=$token")
            }
        }
        return builder.build().toString()
    }
}

File: ios/Yappy/Data/SupportLinks.swift
import Foundation

enum SupportLinks {
    /// Only the case capability is carried onto the configured support site.
    static func url(appeal: Bool = false, source: String? = nil) -> URL {
        var components = URLComponents(string: AppConfig.webUrl + "/support/")!
        let version = (Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String) ?? "unknown"
        let build = (Bundle.main.infoDictionary?["CFBundleVersion"] as? String) ?? "unknown"
        components.queryItems = [URLQueryItem(name: "client", value: "iOS · \(version) (\(build))")]
        if appeal {
            components.queryItems?.append(URLQueryItem(name: "topic", value: "appeal"))
            if let source, let fragment = URLComponents(string: source)?.percentEncodedFragment,
               let token = URLComponents(string: "https://yappy.gg/?" + fragment)?.queryItems?.first(where: { $0.name == "appeal" })?.value,
               !token.isEmpty, token.count <= 2000,
               token.range(of: "^[A-Za-z0-9._-]+$", options: .regularExpression) != nil {
                components.fragment = "appeal=" + token
            }
        }
        return components.url!
    }
}
